lib/SPIRV/SPIRVOps.cpp

```diff
@@ -197,6 +198,144 @@ static void printNoIOOp(Operation *op, OpAsmPrinter *printer) {
+
+  if (parser->parseOperand(compositeInfo) ||
+      parser->getCurrentLocation(&attrLocation) ||
+      parser->parseAttribute(indicesAttr, kIndicesAttrName,
+                             state->attributes) ||
+      parser->parseColonType(compositeType) ||
+      parser->resolveOperand(compositeInfo, compositeType, state->operands)) {
+    return failure();
+  }
+
+  auto indicesArrayAttr = indicesAttr.dyn_cast<ArrayAttr>();
+  if (!indicesArrayAttr) {
+    return parser->emitError(
+        attrLocation,
+        "expected an 32-bit integer array attribute for 'indices'");
+  }
+
+  if (!indicesArrayAttr.size()) {
+    return parser->emitError(attrLocation, "expexted at least one index for ")
+           << compositeType;
+  }
+
+  SmallVector<int32_t, 4> indexes;
+  for (auto indexAttr : indicesArrayAttr) {
+    if (auto indexIntAttr = indexAttr.dyn_cast<IntegerAttr>()) {
+      indexes.push_back(indexIntAttr.getInt());
+    } else {
+      return parser->emitError(attrLocation,
+                               "expexted an 32-bit integer for index");
+    }
+  }
+
+  std::pair<Type, StringRef> resultPair;
+  switch (compositeType.getKind()) {
+  case spirv::TypeKind::Struct:
+  case spirv::TypeKind::Array:
+    resultPair = compositeType.cast<mlir::spirv::CompositeType>()
+                     .getElementTypeByIndices(indexes);
+    if (resultPair.first) {
+      resultType = resultPair.first;
+    } else {
+      return parser->emitError(attrLocation, resultPair.second);
+    }
+    break;
+  case StandardTypes::Vector:
+    resultType = compositeType.cast<VectorType>().getElementType();
+    break;
+  default:
+    return parser->emitError(attrLocation, "invalid type to extract from");
+  }
+
+  if (!resultType) {
+    return parser->emitError(attrLocation,
+                             "can not extract from composite type");
+  }
+
+  state->addTypes(resultType);
+  return success();
+}
+
+static void print(spirv::CompositeExtractOp compositeExtractOp,
+                  OpAsmPrinter *printer) {
+  *printer << spirv::CompositeExtractOp::getOperationName() << ' '
+           << *compositeExtractOp.composite() << compositeExtractOp.indices()
+           << " : " << compositeExtractOp.composite()->getType();
+}
+
+static LogicalResult verify(spirv::CompositeExtractOp compExOp) {
+  auto resultType = compExOp.composite()->getType();
+  auto indicesAttr = compExOp.indices();
+
+  auto indicesArrayAttr = indicesAttr.dyn_cast<ArrayAttr>();
+  if (!indicesArrayAttr) {
+    return compExOp.emitOpError(
+        "expected an 32-bit integer array attribute for 'indices'");
+  }
+
+  if (!indicesArrayAttr.size()) {
+    return compExOp.emitOpError("expexted at least one index for ")
+           << resultType;
+  }
+
+  int32_t index;
+  for (auto indexAttr : indicesArrayAttr) {
+    if (!resultType) {
+      return compExOp.emitError("invalid type to extract from");
+    }
+
+    if (auto indexIntAttr = indexAttr.dyn_cast<IntegerAttr>()) {
```

Comment:
This check is already enforced by `I32ArrayAttr`

Reply:
Thanks, fixed.